core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/retry/RetryPolicyAdapter.java

```diff
@@ -0,0 +1,190 @@
+import software.amazon.awssdk.retries.api.RetryStrategy;
+import software.amazon.awssdk.retries.api.RetryToken;
+import software.amazon.awssdk.retries.api.TokenAcquisitionFailedException;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Implements the {@link RetryStrategy} interface by wrapping a {@link RetryPolicy} instance.
+ */
+@SdkInternalApi
+public final class RetryPolicyAdapter implements RetryStrategy<RetryPolicyAdapter.Builder, RetryPolicyAdapter> {
+
+    private final RetryPolicy retryPolicy;
+    private final RetryPolicyContext retryPolicyContext;
+    private final RateLimitingTokenBucket rateLimitingTokenBucket;
+
+    private RetryPolicyAdapter(Builder builder) {
+        this.retryPolicy = Validate.paramNotNull(builder.retryPolicy, "retryPolicy");
+        this.retryPolicyContext = Validate.paramNotNull(builder.retryPolicyContext, "retryPolicyContext");
+        this.rateLimitingTokenBucket = builder.rateLimitingTokenBucket;
+    }
+
+    @Override
+    public AcquireInitialTokenResponse acquireInitialToken(AcquireInitialTokenRequest request) {
+        RetryPolicyAdapterToken token = new RetryPolicyAdapterToken(request.scope());
+        return AcquireInitialTokenResponse.create(token, rateLimitingTokenAcquire());
+    }
+
+    @Override
+    public RefreshRetryTokenResponse refreshRetryToken(RefreshRetryTokenRequest request) {
+        RetryPolicyAdapterToken token = getToken(request.token());
+        boolean willRetry = retryPolicy.aggregateRetryCondition().shouldRetry(retryPolicyContext);
+        if (!willRetry) {
+            retryPolicy.aggregateRetryCondition().requestWillNotBeRetried(retryPolicyContext);
+            throw new TokenAcquisitionFailedException("Retry policy disallowed retry");
+        }
+        Duration backoffDelay = backoffDelay();
+        return RefreshRetryTokenResponse.create(token, backoffDelay);
+    }
+
+    @Override
+    public RecordSuccessResponse recordSuccess(RecordSuccessRequest request) {
+        RetryPolicyAdapterToken token = getToken(request.token());
+        retryPolicy.aggregateRetryCondition().requestSucceeded(retryPolicyContext);
+        return RecordSuccessResponse.create(token);
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return new Builder(this);
+    }
+
+    RetryPolicyAdapterToken getToken(RetryToken token) {
+        return Validate.isInstanceOf(RetryPolicyAdapterToken.class, token, "Object of class %s was not created by this retry "
+                                                                           + "strategy", token.getClass().getName());
+    }
+
+    boolean isFastFailRateLimiting() {
+        return Boolean.TRUE.equals(retryPolicy.isFastFailRateLimiting());
+    }
+
+    Duration rateLimitingTokenAcquire() {
+        if (!isRateLimitingEnabled()) {
+            return Duration.ZERO;
+        }
+        OptionalDouble tokenAcquireTimeSeconds = rateLimitingTokenBucket.acquireNonBlocking(1.0, isFastFailRateLimiting());
```

Comment:
Can we please add coverage for this case in Junit it currently not shown in coevrage

Reply:
I will add more tests for these after this is merged in a follow up PR; the idea is to make smaller changes to make it easier to review with self-contained minimal changes that build on top of each other.